src/runtime/security/sasl_utils.cpp

```diff
@@ -0,0 +1,156 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "sasl_utils.h"
+#include "kinit_context.h"
+
+#include <sasl/sasl.h>
+#include <sasl/saslplug.h>
+#include <functional>
+
+#include <dsn/c/api_utilities.h>
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/flags.h>
+#include <dsn/utility/synchronize.h>
+
+namespace dsn {
+namespace security {
+DSN_DEFINE_string("security", sasl_plugin_path, "/usr/lib/sasl2", "path to search sasl plugins");
+
+const char *logger_level_to_string(int level)
+{
+    switch (level) {
+    case SASL_LOG_NONE:
+        return "SASL_LOG_NONE";
+    case SASL_LOG_ERR:
+        return "SASL_LOG_ERR";
+    case SASL_LOG_FAIL:
+        return "SASL_LOG_FAIL";
+    case SASL_LOG_WARN:
+        return "SASL_LOG_WARN";
+    case SASL_LOG_NOTE:
+        return "SASL_LOG_NOTE";
+    case SASL_LOG_DEBUG:
+        return "SASL_LOG_DEBUG";
+    case SASL_LOG_TRACE:
+        return "SASL_LOG_TRACE";
+    case SASL_LOG_PASS:
+        return "SASL_LOG_PASS";
+    default:
+        return "Unkown SASL log level";
+    }
+}
+
+int sasl_simple_logger(void *context, int level, const char *msg)
+{
+    if (SASL_LOG_NONE == level || nullptr == msg) {
+        return SASL_OK;
+    }
+
+    ddebug_f("sasl log info: log level = {}, message = {}", logger_level_to_string(level), msg);
+    return SASL_OK;
+}
+
+int sasl_get_path(void *context, char **path)
+{
+    if (nullptr == path) {
+        return SASL_BADPARAM;
+    }
+    *path = const_cast<char *>(FLAGS_sasl_plugin_path);
+    return SASL_OK;
+}
+
+int sasl_get_username(void *context, int id, const char **result, unsigned *len)
+{
+    if (nullptr == result) {
+        return SASL_BADPARAM;
+    }
+    static const std::string username = get_username();
+    switch (id) {
+    case SASL_CB_USER:
+    case SASL_CB_AUTHNAME:
+        *result = username.c_str();
+        if (len != nullptr) {
+            *len = username.length();
+        }
+        return SASL_OK;
+    default:
+        dassert_f(false, "unexpected SASL callback type: {}", id);
+        return SASL_BADPARAM;
+    }
+}
+
+sasl_callback_t client_callbacks[] = {
+    {SASL_CB_USER, (sasl_callback_ft)&sasl_get_username, nullptr},
+    {SASL_CB_GETPATH, (sasl_callback_ft)&sasl_get_path, nullptr},
+    {SASL_CB_AUTHNAME, (sasl_callback_ft)&sasl_get_username, nullptr},
+    {SASL_CB_LOG, (sasl_callback_ft)&sasl_simple_logger, nullptr},
+    {SASL_CB_LIST_END, nullptr, nullptr}};
```

Comment:
What the meaning of `LIST_END` in `SASL_CB_LIST_END`?

Reply:
it means the end of the list. It just used for a smybol to show it is the end of the list.